File: go/rendezvous.go
package consistenthash

import (
	"math"
	"sort"
)

// Rendezvous is highest-random-weight (HRW) placement: score every node for the key
// and take the max. No ring to rebuild, and weights are first-class. Lookups are
// O(N), which suits small clusters where that is cheap.
type Rendezvous struct {
	weights map[string]int
}

// NewRendezvous builds an unweighted HRW placer over nodes.
func NewRendezvous(nodes []string) *Rendezvous {
	h := &Rendezvous{weights: make(map[string]int)}
	for _, n := range nodes {
		h.Add(n, 1)
	}
	return h
}

// Add registers a node with an integer weight (>= 1). A heavier node wins more keys.
func (h *Rendezvous) Add(node string, weight int) {
	if weight < 1 {
		panic("weight must be at least 1")
	}
	h.weights[node] = weight
}

// Remove drops a node.
func (h *Rendezvous) Remove(node string) { delete(h.weights, node) }

// Get returns the node with the highest score for key, or "" if empty.
func (h *Rendezvous) Get(key string) string {
	if len(h.weights) == 0 {
		return ""
	}
	weighted := h.isWeighted()
	best := ""
	bestScore := math.Inf(-1)
	for node, w := range h.weights {
		s := h.score(node, key, w, weighted)
		if s > bestScore || (s == bestScore && node > best) {
			bestScore = s
			best = node
		}
	}
	return best
}

// GetReplicas returns the top count nodes by score.
func (h *Rendezvous) GetReplicas(key string, count int) []string {
	if len(h.weights) == 0 || count <= 0 {
		return []string{}
	}
	weighted := h.isWeighted()
	type scored struct {
		node  string
		score float64
	}
	ranked := make([]scored, 0, len(h.weights))
	for node, w := range h.weights {
		ranked = append(ranked, scored{node, h.score(node, key, w, weighted)})
	}
	sort.Slice(ranked, func(i, j int) bool {
		if ranked[i].score != ranked[j].score {
			return ranked[i].score > ranked[j].score
		}
		return ranked[i].node > ranked[j].node
	})
	cap := count
	if cap > len(ranked) {
		cap = len(ranked)
	}
	out := make([]string, cap)
	for i := 0; i < cap; i++ {
		out[i] = ranked[i].node
	}
	return out
}

// Nodes returns the nodes in sorted order.
func (h *Rendezvous) Nodes() []string {
	out := make([]string, 0, len(h.weights))
	for n := range h.weights {
		out = append(out, n)
	}
	sort.Strings(out)
	return out
}

// Len is the number of nodes.
func (h *Rendezvous) Len() int { return len(h.weights) }

func (h *Rendezvous) isWeighted() bool {
	for _, w := range h.weights {
		if w != 1 {
			return true
		}
	}
	return false
}

func (h *Rendezvous) score(node, key string, weight int, weighted bool) float64 {
	hv := Fnv1a64(node + "#" + key)
	if !weighted {
		return float64(hv) // pure integer score - portable, what conformance pins
	}
	normalized := (float64(hv) + 1) / (math.MaxUint64 + 1.0)
	return -float64(weight) / math.Log(normalized)
}

File: csharp/Rendezvous.cs
namespace ConsistentHash;

/// <summary>
/// Highest-random-weight (rendezvous) placement: score every node for the key and
/// take the max. No ring to rebuild, weights are first-class, lookups are O(N).
/// </summary>
public sealed class Rendezvous
{
    private readonly Dictionary<string, int> _weights = new();

    public Rendezvous(IEnumerable<string>? nodes = null)
    {
        if (nodes is not null)
        {
            foreach (var n in nodes)
            {
                Add(n);
            }
        }
    }

    public void Add(string node, int weight = 1)
    {
        if (weight < 1)
        {
            throw new ArgumentOutOfRangeException(nameof(weight), "weight must be at least 1");
        }

        _weights[node] = weight;
    }

    public void Remove(string node) => _weights.Remove(node);

    public string? Get(string key)
    {
        if (_weights.Count == 0)
        {
            return null;
        }

        var weighted = IsWeighted();
        string? best = null;
        var bestScore = double.NegativeInfinity;
        foreach (var (node, weight) in _weights)
        {
            var s = Score(node, key, weight, weighted);
            if (s > bestScore || (s == bestScore && string.CompareOrdinal(node, best) > 0))
            {
                bestScore = s;
                best = node;
            }
        }

        return best;
    }

    public IReadOnlyList<string> GetReplicas(string key, int count)
    {
        if (_weights.Count == 0 || count <= 0)
        {
            return Array.Empty<string>();
        }

        var weighted = IsWeighted();
        return _weights
            .Select(kv => (node: kv.Key, score: Score(kv.Key, key, kv.Value, weighted)))
            .OrderByDescending(x => x.score)
            .ThenByDescending(x => x.node, StringComparer.Ordinal)
            .Take(Math.Min(count, _weights.Count))
            .Select(x => x.node)
            .ToList();
    }

    public IReadOnlyList<string> Nodes =>
        _weights.Keys.OrderBy(n => n, StringComparer.Ordinal).ToList();

    public int Count => _weights.Count;

    private bool IsWeighted() => _weights.Values.Any(w => w != 1);

    private static double Score(string node, string key, int weight, bool weighted)
    {
        var h = HashRing.Fnv1a64($"{node}#{key}");
        if (!weighted)
        {
            return h; // pure integer score - portable, what conformance pins
        }

        var normalized = ((double)h + 1) / ((double)ulong.MaxValue + 1);
        return -weight / Math.Log(normalized);
    }
}
